Input component

drillprop opened the pull request:
Fix #2707 
Designs: https://www.figma.com/file/Pf31tuYpozYmpq163U2ho8/Web-components?node-id=7058%3A174881

drillprop committed:
update name

drillprop committed:
update input component

drillprop committed:
create inputStyles.utils, update stories

drillprop committed:
lint

drillprop committed:
fix padding

toiletgranny reviewed (changes requested):
@drillprop Marvelous! I especially like how it looks out in the wild, really breathing some fresh air into some of the views in the app. <3

Here's what I found:
1. Both icons and the spinner in the processing state are not dimmed in the Input disabled state.
2. Having set the `type` to `number` in Storybook, I couldn't get the numbers to be formatted [as described in Figma docs](https://www.figma.com/file/Pf31tuYpozYmpq163U2ho8/Web-components?node-id=7058%3A174881) (by the way, please disregard commas as thousand separators, which I already changed to spaces). Was that not possible to implement?
3. The Button component is placed a bit too far away from the right edge of the Input component. If you give it another look in Figma, you'll notice that buttons are actually moved a bit closer to the edge, to accommodate for the additional padding they come with. This is so that when the Button component is not being hovered, its icon is positioned in the same way as if it _was_ just an icon. Happy to jump on a Slack call to discuss this further.
4. I can activate the Input component and then the Button component by tabbing — I respect that. But, could we preserve the focus styling of the input for when the Button component (or any other component) inside is focused? I think this is what `:focus-within` might help us with?
5. Unfortunately, I could not review the styling of prefixes and suffixes, as well as other custom components that could be potentially embedded in the Input (such as a small Pill component, or a 24px Avatar component). Can we have those as props in the Storybook?

On top of that, I went for a quick tour through the app to see the Input component in real-life. Accidentally, I've stumbled upon this issue, where the "License, content rating, published before, marketing, delete video" description was placed next to the button component, instead of underneath it. Not sure if this is because of this PR, but I haven't seen this before. Either way, @kdembler, could it be yet another argument for us to consider introducing automated visual regression tests? 🤔
<img width="651" alt="CleanShot 2022-06-06 at 11 01 32@2x" src="https://user-images.githubusercontent.com/35307309/172130370-422af94e-e295-482d-aba2-46c2375d41d9.png">

drillprop commented:
> 2. Having set the `type` to `number` in Storybook, I couldn't get the numbers to be formatted [as described in Figma docs](https://www.figma.com/file/Pf31tuYpozYmpq163U2ho8/Web-components?node-id=7058%3A174881) (by the way, please disregard commas as thousand separators, which I already changed to spaces). Was that not possible to implement?

I won't be implementing separators in this PR. This will be tackled here https://github.com/Joystream/atlas/issues/2595

drillprop committed:
fix focus, fix stories

drillprop commented:
> 1. Both icons and the spinner in the processing state are not dimmed in the Input disabled state.

Fixed

> 3. The Button component is placed a bit too far away from the right edge of the Input component. If you give it another look in Figma, you'll notice that buttons are actually moved a bit closer to the edge, to accommodate for the additional padding they come with. This is so that when the Button component is not being hovered, its icon is positioned in the same way as if it _was_ just an icon. Happy to jump on a Slack call to discuss this further.

Should be fixed

> 4. I can activate the Input component and then the Button component by tabbing — I respect that. But, could we preserve the focus styling of the input for when the Button component (or any other component) inside is focused? I think this is what `:focus-within` might help us with?

Fixed

> 5. Unfortunately, I could not review the styling of prefixes and suffixes, as well as other custom components that could be potentially embedded in the Input (such as a small Pill component, or a 24px Avatar component). Can we have those as props in the Storybook?

Fixed. Unfortunately, you won't be able to play around a lot. I had some issues with setting this right in the storybook. I just add a few examples with different elements as suffixes and prefixes. 

> On top of that, I went for a quick tour through the app to see the Input component in real-life. Accidentally, I've stumbled upon this issue, where the "License, content rating, published before, marketing, delete video" description was placed next to the button component, instead of underneath it. Not sure if this is because of this PR, but I haven't seen this before. Either way, @kdembler, could it be yet another argument for us to consider introducing automated visual regression tests? 🤔 <img alt="CleanShot 2022-06-06 at 11 01 32@2x" width="651" src="https://user-images.githubusercontent.com/35307309/172130370-422af94e-e295-482d-aba2-46c2375d41d9.png">

We discussed this yesterday. I think this came from the latest form field update.  I'll tackle this in another PR

drillprop committed:
fix license issue

drillprop commented:
> We discussed this yesterday. I think this came from the latest form field update. I'll tackle this in another PR

I was wrong actually - I introduced the bug in this PR 😅 . Good catch - should be fixed now.

kdembler commented on packages/atlas/src/components/_inputs/Input/Input.stories.tsx:
```diff
@@ -0,0 +1,124 @@
+  return <Input {...args} onChange={(e) => setValue(e.currentTarget.value)} value={value} />
+}
+
+const TemplateWithUncontrolledInput: Story<InputProps> = (args) => {
```
Maybe we just make this template the default? (Both are uncontrolled)

kdembler commented on packages/atlas/src/components/_inputs/inputs.utils.ts:
```diff
@@ -0,0 +1,88 @@
+  large: sizes(3),
+} as const
+
+export const getPadding = (size: InputSize, nodeWidthProps?: NodeWidthProps) => {
```
`getInputPadding` for consistency?

kdembler commented on packages/atlas/src/components/_inputs/inputs.utils.ts:
```diff
@@ -0,0 +1,88 @@
+}
+
+export const getInputPseudoSelectorStyles = ({ error }: { error?: boolean }) => css`
+  ${sharedInputStyles.default};
```
This function will also apply the default styles, wich are not pseudo selectors so the name may be slightly misleading. Maybe `getBaseInputStyles`?

kdembler commented on packages/atlas/src/components/_inputs/Input/Input.styles.ts:
```diff
@@ -0,0 +1,94 @@
+
+export const TextInput = styled.input<TextInputProps>`
+  width: 100%;
+  padding: ${sizes(3)} ${sizes(5)};
```
Won't this be overwritten by `getPadding` anyway?

kdembler commented on packages/atlas/src/components/_inputs/Input/Input.styles.ts:
```diff
@@ -0,0 +1,94 @@
+
+  ${({ inputSize, leftNodeWidth, rightNodeWidth }) => getPadding(inputSize, { leftNodeWidth, rightNodeWidth })};
+
+  ${({ error }) => getInputPseudoSelectorStyles({ error })};
```
If you wanted to, this could be just `${getInputPseudoSelectorStyles};` Similarly, if you updated the args of `getPadding` to accept a single object, it could be used the same way. But leaving this up to you, maybe it's better to be explicit

drillprop replied:
Updated after all

kdembler commented on packages/atlas/src/components/_inputs/Input/Input.styles.ts:
```diff
@@ -0,0 +1,94 @@
+  left
+    ? css`
+        left: 0;
+        padding-left: calc(${horizontalPadding[size]});
```
Unneeded `calc` I think

kdembler commented on packages/atlas/src/components/_inputs/Input/Input.styles.ts:
```diff
@@ -0,0 +1,94 @@
+      `
+    : css`
+        right: 0;
+        padding-right: calc(${horizontalPadding[size]} - ${isButton ? sizes(1) : '0px'});
```
Maybe better to have the condition outside of `calc` so we don't do `calc(100px - 0px)`?

kdembler commented on packages/atlas/src/components/_inputs/Input/Input.stories.tsx:
```diff
@@ -0,0 +1,124 @@
+        nodeEnd={
+          <Text secondary color={cVar('colorTextMuted')} variant="t300">
+            $
+          </Text>
```
I'm not sure if this is captured in the Figma docs but when we discussed the suffix/prefix with Adam, he mentioned that it would be good for all those basic text nodes to always use the same styling.

What do you think about this: We could check whether `nodeStart` (or end) is a string and if it is, we could wrap it in this `<Text>` inside the `Input` itself, not requiring the consumers to repeat the style?

Btw, I think you don't need both `secondary` and `color` props

drillprop replied:
good idea - added

kdembler commented on packages/atlas/src/components/_inputs/Input/Input.tsx:
```diff
@@ -0,0 +1,126 @@
+        defaultValue={defaultValue}
+      />
+      {nodeStart && (
+        <NodeContainer onClick={() => inputRef.current?.focus()} size={size} ref={nodeLeftRef} left disabled={disabled}>
```
We didn't need this `onClick` logic for the old `TextField` and I think it worked. What changed?

drillprop replied:
The old `Textfield` was wrapped in `label`(`InputBase`). That's why we didn't have to add this logic.

drillprop committed:
CR fixes

drillprop committed:
fix placeholder for firefox, use Text inside Input

kdembler commented on packages/atlas/src/components/_inputs/Input/Input.tsx:
```diff
@@ -61,6 +63,24 @@ const InputComponent: React.ForwardRefRenderFunction<HTMLInputElement, InputProp
+    </Text>
+  ) : (
+    nodeEnd
+  )
```
Let's rename those vars - `render*` suggests it's a function that returns JSX while it's a static JSX instead. Maybe `rendered`?

drillprop committed:
tweaks

drillprop committed:
bring back onClick and div as container

drillprop committed:
remove onClick handler
